File: 2015/rust/src/day1.rs
use simple_error::{bail, SimpleError};
use std::error::Error;
use std::fs::File;
use std::io::prelude::*;
use std::time::Instant;

pub fn day1(input_file: &str) -> Result<(), Box<dyn Error>> {
    println!("Day 1: Not Quite Lisp");

    let mut f = File::open(input_file)?;
    let mut input = String::new();
    f.read_to_string(&mut input)?;

    let start = Instant::now();
    println!("\tPart 1: {} ({:?})", part1(&input)?, start.elapsed());
    let second = Instant::now();
    println!("\tPart 2: {} ({:?})", part2(&input)?, second.elapsed());
    println!("\t\t Completed in {:?}", start.elapsed());
    Ok(())
}

fn part1(input: &str) -> Result<i32, Box<dyn Error>> {
    input.chars().try_fold(0, |floor, c| match c {
        '(' => Ok(floor + 1),
        ')' => Ok(floor - 1),
        _ => bail!("Invalid input: {}", c),
    })
}

#[derive(Debug)]
enum Basement {
    Floor(i32),
    Err(Box<dyn Error>),
}
impl std::error::Error for Basement {}
impl std::fmt::Display for Basement {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "")
    }
}

fn part2(input: &str) -> Result<i32, Box<dyn Error>> {
    match input
        .chars()
        .enumerate()
        .try_fold(0, |mut floor, (step, c)| match c {
            '(' => Ok(floor + 1),
            ')' => {
                floor -= 1;
                if floor < 0 {
                    Err(Basement::Floor(step as i32))
                } else {
                    Ok(floor)
                }
            }
            _ => Err(Basement::Err(SimpleError::new("Invalid input").into())),
        }) {
        Err(Basement::Floor(step)) => Ok(step),
        _ => bail!("Solution not found"),
    }
}

File: 2015/c/src/day1.c
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include "lib.h"

static int part1(const char *input, size_t input_len)
{
    int floor = 0;
    for (int i = 0; i < input_len; i++)
    {
        switch (input[i])
        {
        case '(':
            floor++;
            break;
        case ')':
            floor--;
            break;
        default:
            errno = EINVAL;
            return -1;
        }
    }

    return floor;
}

static int part2(const char *input, size_t input_len)
{
    int floor = 0;
    for (int i = 0; i < input_len; i++)
    {
        switch (input[i])
        {
        case '(':
            floor++;
            break;
        case ')':
            floor--;
            break;
        default:
            errno = EINVAL;
            return -1;
        }

        if (floor == -1)
            return i + 1;
    }
    return -EINVAL;
}

int day1(const char *filename)
{
    printf("Day 1: Not Quite Lisp\n");

    char *input;
    int filesize = read_file_to_buffer(&input, filename);
    if (filesize < 0)
        return -1;

    int rval = part1(input, filesize);
    if (rval < 0)
        return rval;
    printf("\tPart 1: %d\n", rval);

    rval = part2(input, filesize);
    if (rval < 0)
        return rval;
    printf("\tPart 2: %d\n", rval);

    free(input);

    return 0;
}
